### PR TITLE
feat: refresh-token

### PR DESCRIPTION
**Pull request by WTW0313**
# Checklist:

> [!IMPORTANT]  
> Please review the checklist below before submitting your pull request.

- [ ] Please open an issue before creating a PR or link to an existing issue
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I ran `dev/reformat`(backend) and `cd web && npx lint-staged`(frontend) to appease the lint gods

# Description

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue. Close issue syntax: `Fixes #<issue number>`, see [documentation](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) for more details.

Fixes 

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [ ] Improvement, including but not limited to code refactoring, performance optimization, and UI/UX improvement
- [ ] Dependency upgrade

# Testing Instructions

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

### CODE REVIEW

File: web/yarn.lock
Comment (crazywoola) [inline web/yarn.lock:1]: This should not be committed into this pr.

File: web/hooks/use-refresh-token.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+  const interval = useRef<number>(55 * 60 * 1000)
+
+  const getExpireTime = useCallback((token: string) => {
+    if (!token)
+      return 0
+    const decoded = jwtDecode(token)
+    return (decoded.exp || 0) * 1000
+  }, [])
+
+  const getCurrentTimeStamp = useCallback(() => {
```
Comment (crazywoola) [inline web/hooks/use-refresh-token.ts:26]:
```suggestion
  const getCurrentTimestamp = useCallback(() => {
```

File: web/hooks/use-refresh-token.ts [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    router.replace('/signin')
+  }, [])
+
+  const getNewAccessToken = useCallback(async (currentAccessToken: string, currentRefreshToken: string) => {
+    if (localStorage?.getItem('is_refreshing') === '1')
+      return null
+    const currentTokenExpireTime = getExpireTime(currentAccessToken)
+    let lastRefreshTime = parseInt(localStorage?.getItem('last_refresh_time') || '0')
+    lastRefreshTime = isNaN(lastRefreshTime) ? 0 : lastRefreshTime
+    if (getCurrentTimeStamp() + advanceTime.current > currentTokenExpireTime
```
Comment (crazywoola) [inline web/hooks/use-refresh-token.ts:44]:
```suggestion
    if (getCurrentTimestamp() + advanceTime.current > currentTokenExpireTime
```

File: web/hooks/use-refresh-token.ts [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+      && lastRefreshTime + interval.current < getCurrentTimeStamp()) {
```
Comment (crazywoola) [inline web/hooks/use-refresh-token.ts:45]:
```suggestion
      && lastRefreshTime + interval.current < getCurrentTimestamp()) {
```

**Review by iamjoel — APPROVED** [review]

**Pull request merged**